File: typescript/src/mcp_code_mode_loop.ts
/**
 * TypeScript MCP Code Mode loop: hosted search_docs → LLM plan → hosted execute.
 * Mirrors agents/mcp_code_mode_loop.py with local SDK fallback.
 */
import ContextDev from "context.dev";
import { planMcpExecute } from "./llm_policy.js";
import { buildExecuteTypescript } from "./mcp_execute_codegen.js";
import { OP_CREDITS, normalizeExecutePlan } from "./mcp_op_map.js";
import {
  extractDomain,
  hostedExecute,
  hostedSearchDocs,
} from "./mcp_client.js";
import type { ExecuteStep } from "./llm_policy.js";

function getClient(): ContextDev {
  const key = process.env.CONTEXT_DEV_API_KEY ?? process.env.CONTEXT_API_KEY;
  if (!key) throw new Error("CONTEXT_DEV_API_KEY is required");
  return new ContextDev({ apiKey: key });
}

async function withRetry<T>(fn: () => Promise<T>): Promise<T> {
  for (let i = 0; i < 3; i++) {
    try {
      return await fn();
    } catch (e: unknown) {
      const status = (e as { status?: number }).status;
      if ((status === 408 || status === 429) && i < 2) {
        await new Promise((r) => setTimeout(r, 1000 * 2 ** i));
        continue;
      }
      throw e;
    }
  }
  throw new Error("retry exhausted");
}

function isHostedExecuteFailure(raw: Record<string, unknown>): boolean {
  if (typeof raw.raw === "string") {
    const t = raw.raw;
    return t.includes("TS Diagnostics") || t.toLowerCase().includes("error");
  }
  const result = raw.result as Record<string, unknown> | undefined;
  const identity = result?.identity as { title?: string } | undefined;
  return !identity?.title;
}

async function executePlanLocal(
  plan: ExecuteStep[]
): Promise<{ result: Record<string, unknown>; opsRun: string[]; credits: number }> {
  const client = getClient();
  const out: Record<string, unknown> = {};
  const opsRun: string[] = [];
  let credits = 0;

  for (const step of plan) {
    const domain = step.args.domain;
    if (!domain) continue;

    if (step.op === "brand.retrieve") {
      const brandRes = await withRetry(() => client.brand.retrieve({ domain }));
      out.identity = {
        title: brandRes.brand?.title ?? null,
        logo_url: brandRes.brand?.logos?.[0]?.url ?? null,
      };
      credits += OP_CREDITS["brand.retrieve"];
      opsRun.push(`local:brand.retrieve(${domain})`);
    } else if (step.op === "web.extract_styleguide") {
      const styleRes = await withRetry(() => client.brand.styleguide({ domain }));
      const colors = styleRes.styleguide?.colors;
      out.design_tokens = {
        color_count: [colors?.accent, colors?.background, colors?.text].filter(Boolean).length,
        has_typography: !!styleRes.styleguide?.typography,
      };
      credits += OP_CREDITS["web.extract_styleguide"];
      opsRun.push(`local:brand.styleguide(${domain})`);
    } else if (step.op === "web.scrape_sitemap") {
      const smRes = await withRetry(() => client.brand.webScrapeSitemap({ domain }));
      out.site_scale = {
        url_count: (smRes.urls ?? []).length,
        sample_paths: (smRes.urls ?? []).slice(0, 3),
      };
      credits += OP_CREDITS["web.scrape_sitemap"];
      opsRun.push(`local:brand.webScrapeSitemap(${domain})`);
    }
  }

  return { result: { ...out, credits_estimated: credits }, opsRun, credits };
}

export async function runCodeModeLoop(goal: string): Promise<Record<string, unknown>> {
  const domain = extractDomain(goal);
  const hits = await hostedSearchDocs(goal);
  const { plan: rawPlan, policySource } = await planMcpExecute(goal, hits);
  const planSteps = normalizeExecutePlan(goal, rawPlan, hits, domain);
  const code = buildExecuteTypescript(domain, planSteps);

  let executeSource = "hosted_mcp_execute";
  let result: Record<string, unknown>;
  let opsRun: string[];
  let credits = planSteps.reduce((sum, s) => sum + (OP_CREDITS[s.op] ?? 0), 0);

  try {
    const raw = await hostedExecute(code, goal);
    if (isHostedExecuteFailure(raw)) {
      throw new Error(
        typeof raw.raw === "string" ? raw.raw.slice(0, 200) : "hosted execute empty identity"
      );
    }
    const payload =
      typeof raw.result === "object" && raw.result !== null
        ? (raw.result as Record<string, unknown>)
        : raw;
    result = { ...payload, credits_estimated: credits };
    opsRun = planSteps.map((s) => `hosted:execute(${s.op})`);
  } catch (exc) {
    const local = await executePlanLocal(planSteps);
    result = local.result;
    opsRun = local.opsRun;
    credits = local.credits;
    executeSource = `local_sdk_fallback:${exc instanceof Error ? exc.constructor.name : "Error"}`;
  }

  return {
    goal,
    search_docs: { query: goal, source: "hosted_mcp", hits },
    execute_plan: planSteps,
    policy_source: policySource,
    execute_source: executeSource,
    execute_typescript: code,
    result,
    ops_run: opsRun,
    credits_estimated: credits,
  };
}

async function main(): Promise<void> {
  const goal =
    process.argv.slice(2).join(" ") ||
    "Get stripe.com brand identity, design tokens, and site scale";
  console.log(JSON.stringify(await runCodeModeLoop(goal), null, 2));
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: agents/mcp_code_mode_loop.py
#!/usr/bin/env python3
"""
MCP Code Mode agent loop: search_docs → LLM execute plan → run → observe.

Mirrors hosted MCP at context-dev.stlmcp.com (search_docs + execute tools).

Run:
  CONTEXT_DEV_API_KEY=... OPEN_ROUTER_KEY=... \\
    python agents/mcp_code_mode_loop.py "Get stripe.com brand, design tokens, and site scale"
"""

from __future__ import annotations

import json
import sys
from pathlib import Path
from typing import Any

ROOT = Path(__file__).resolve().parent.parent
sys.path.insert(0, str(ROOT))

from agents.llm_policy import plan_mcp_execute  # noqa: E402
from src.context_client import (  # noqa: E402
    create_client,
    extract_styleguide,
    retrieve_brand,
    scrape_sitemap,
)

DOC_INDEX = Path(__file__).resolve().parent / "sdk_doc_index.json"


def search_docs(query: str, limit: int = 3) -> list[dict[str, str]]:
    """MCP search_docs analogue — keyword search over local SDK index."""
    entries = json.loads(DOC_INDEX.read_text())
    q = query.lower()
    scored: list[tuple[int, dict]] = []
    for entry in entries:
        score = sum(1 for kw in entry["keywords"] if kw in q)
        if score:
            scored.append((score, entry))
    scored.sort(key=lambda x: -x[0])
    hits = [e for _, e in scored[:limit]]
    if not hits:
        hits = entries[:limit]
    return [
        {
            "id": h["id"],
            "signature": h["signature"],
            "returns": h["returns"],
            "credits": str(h["credits"]),
        }
        for h in hits
    ]


def execute_plan(plan: list[dict[str, Any]]) -> tuple[dict[str, Any], list[str], int]:
    """MCP execute analogue — run composed SDK operations, return filtered JSON."""
    client = create_client()
    ops_run: list[str] = []
    credits = 0
    out: dict[str, Any] = {}

    for step in plan:
        op = step.get("op", "")
        args = step.get("args") or {}
        domain = args.get("domain", "")

        if op == "brand.retrieve":
            brand = retrieve_brand(client, domain)
            out["identity"] = {
                "title": brand.get("title"),
                "logo_url": brand.get("logo_url"),
            }
            credits += 10
            ops_run.append(f"brand.retrieve({domain})")
        elif op == "web.extract_styleguide":
            style = extract_styleguide(client, domain)
            out["design_tokens"] = {
                "color_count": style.get("color_count"),
                "has_typography": style.get("has_typography"),
            }
            credits += 10
            ops_run.append(f"web.extract_styleguide({domain})")
        elif op in ("web.scrape_sitemap", "web.web_scrape_sitemap"):
            sm = scrape_sitemap(client, domain)
            out["site_scale"] = {
                "url_count": sm.get("url_count"),
                "sample_paths": sm.get("sample_urls"),
            }
            credits += 1
            ops_run.append(f"web.scrape_sitemap({domain})")

    out["credits_estimated"] = credits
    return out, ops_run, credits


def run_code_mode_loop(goal: str) -> dict[str, Any]:
    doc_query = goal if "brand" in goal.lower() else f"brand styleguide sitemap {goal}"
    doc_hits = search_docs(doc_query)
    execute_plan_steps, policy_source = plan_mcp_execute(goal, doc_hits)
    result, ops_run, credits = execute_plan(plan_steps := execute_plan_steps)

    return {
        "goal": goal,
        "search_docs": {"query": doc_query, "hits": doc_hits},
        "execute_plan": plan_steps,
        "policy_source": policy_source,
        "result": result,
        "ops_run": ops_run,
        "credits_estimated": credits,
    }


def main() -> int:
    goal = (
        " ".join(sys.argv[1:])
        if len(sys.argv) > 1
        else "Get stripe.com brand identity, design tokens, and site scale"
    )
    print(json.dumps(run_code_mode_loop(goal), indent=2))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())